discord_bot/discord_bot.py

```diff
@@ -1,60 +1,86 @@
-        r = requests.get(DISCORD_GATEWAY_ENDPOINT)
+    def get_gateway(self):
+        """
+        Caches a gateway value and retrieves a new URL
+        :return: gateway URL
+        """
+        r = requests.get(self.config["discord_gateway_endpoint"])
         return r.json()['url']
 
-    async def handshake(self, websocket):
-        identity_obj = {
-            "token": self.config["bot_token"],
-            "properties": {
-                "$os": "windows",
-                "$browser": "disco",
-                "$device": "disco"
-            },
-            "compress": False,
-            "large_threshold": 250,
-            "shard": [1, 10]
-        }
-        await websocket.send(json.dumps({"op": 2, "d": identity_obj}))
+    async def send_json(self, payload):
+        asyncio.ensure_future(self.websocket.send(json.dumps(payload)))
 
-    async def gateway_handler(self):
+    async def handshake(self, message):
+        """
+        When a websocket connection is opened, Hello payload is received. Then, an Identify payload is sent as
+        part of the handshake to authorize this client.
 
+        :param message: dict - hello payload
+        """
+        self.heartbeat_interval_ms = message["d"]["heartbeat_interval"]
+        asyncio.ensure_future(self.heartbeat())
+
+        handshake_identity = self.config["handshake_identity"]
+        asyncio.ensure_future(self.send_json({"op": Opcodes.IDENTIFY, "d": handshake_identity}))
+
+    async def heartbeat(self):
+        """
+        Sends a heartbeat payload every heartbeat interval.
+        """
+        await asyncio.sleep(self.heartbeat_interval_ms / 1000.0)
+        asyncio.ensure_future(self.send_json({"op": Opcodes.HEARTBEAT, "d": self.last_seq}))
+        asyncio.ensure_future(self.heartbeat())
+
+    async def gateway_handler(self):
+        """
+        Creates the websocket, receives messages and acts on them.
+        """
         async with websockets.connect("{}/?v={}&encoding={}".format(self.gateway_ws_url,
                                                                     self.config["gateway_api_version"],
                                                                     self.config["gateway_encoding"])) as websocket:
+            self.websocket = websocket
             while True:
-                message = await websocket.recv()
-                print(message)
+                message = await self.websocket.recv()
                 message = json.loads(message)
-                if message["op"] == 10:
-                    await self.handshake(websocket)
-
-
-# main entry point
-if __name__ == "__main__":
-    m = DiscordBot()
+                print("{}: {}".format(Opcodes(message["op"]).name, message))
+                if message["op"] == Opcodes.HELLO:
+                    asyncio.ensure_future(self.handshake(message))
+                elif message["op"] == Opcodes.DISPATCH:
+                    pass
+                elif not message["s"]:
```

Comment:
This is definitely not what we want to do. We NEVER want to reset `self.last_seq` to `None` since the Discord spec says:

> The inner d key must be set to the last seq (s) received by the client. If none has yet been received you should send null (you cannot send a heartbeat before authenticating, however).

This means that we need to keep track of the newest, non-null value for last seq.